### ISSUE TITLE
Rook begins terminating next OSD before current OSD pod is running

### ISSUE THREAD

**Issue by BlaineEXE** [rook/rook#2181]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

Output from `kubectl get pods` below (some pods removed to reduce output) shows that while osd-2's pod is still in `Init` stage, Rook has already begun terminating osd-6. In a cluster with a replication factor of 2, this could easily result in data loss unless there is some health checking going on during termination. Even if data is safe, I think it may be unnerving for users to see Rook seemingly reporting that it's taking more osds offline than is safe.
```
blaine-dev-caasp-admin:~ # kubectl get pods
NAME                                   READY     STATUS             RESTARTS   AGE
...
rook-ceph-osd-0-5f54c97c8c-cx2hp       1/1       Running            0          1m
rook-ceph-osd-2-64d9588dcd-85rxb       0/1       Init:0/1           0          2s
rook-ceph-osd-id-1-6bd6cf7f4f-gjw59    1/1       Running            0          19m
rook-ceph-osd-id-2-66d6c944ff-m2jpz    0/1       Terminating        0          19m
rook-ceph-osd-id-3-7b87f8dcd5-h2c26    1/1       Running            0          18m
rook-ceph-osd-id-4-7d765f857b-j6bcv    1/1       Running            0          19m
rook-ceph-osd-id-5-7d5bb7f59b-pffmj    1/1       Running            0          18m
rook-ceph-osd-id-6-6d5b64b74f-d9fp9    1/1       Terminating        0          19m
rook-ceph-osd-id-7-5c7bbdf8bf-ddwxs    1/1       Running            0          18m
```

**Expected behavior:**
To make sure my cluster was updating safely, I think I would expect to see all osds `Running` before seeing another osd enter the `Terminating` state.

**How to reproduce it (minimal and precise):**
1. Install Rook v0.8.2
2. Update to the master image `kubectl -n rook-ceph-system set image deploy/rook-ceph-operator rook-ceph-operator=<my-custom-registry>:5000/rook-ceph:latest`
3. Observe osd upgrade behavior

**Environment**:
* OS: CaaSP's OS is a SLE 12 SP3-based OS
* Kernel: `4.4.132-94.33-default #1 SMP Tue May 29 20:09:56 UTC 2018 (76aae3b) x86_64 x86_64 x86_64 GNU/Linux`
* Cloud provider: OpenStack
* Rook version: N/A
* Kubernetes version: 1.9.8
* Kubernetes cluster type: CaaSP 3.0
* Storage backend status: N/A

**Comment by BlaineEXE** [comment]
@travisn I observed this behavior while testing my mds upgrade. I haven't poked at the code that defines this, so I'm not certain that there is any danger of rook actually bringing 2 osds down at the same time. If it is, that might be a problem we need to fix soon.

If it's not, it's less of an issue, but I think from a user standpoint, they would rather see the behavior I was expecting, which wouldn't likely flag a false-positive type estimation of safety at first glance. In this case I'd leave changing this behavior for the future.

**Comment by travisn** [comment]
@BlaineEXE Do you still have the operator logs? If so, do you see messages like the following in the log? 
```gave up waiting for deployment rook-ceph-osd-0 to update"```

In [this function](https://github.com/rook/rook/blob/master/pkg/operator/k8sutil/deployment.go#L89), the operator will wait up to two minutes for an osd deployment to be updated. If it does not update in that amount of time, it will log an error and continue with the next OSD. This may be too aggressive of a timeout, but it didn't seem appropriate to wait indefinitely either. Suggestions?

**Comment by BlaineEXE** [comment]
I do see some of those in the log, yes. As far as the osds go, I'm not sure if I can say whether that value is reasonable or not. As I am watching the most recent upgrade happen now, it seems that the behavior may be that the old osd doesn't actually finish terminating until the new one is initialized. Maybe what needs to happen is to terminate the old/current osd and then start bringing the new one up immediately, then wait for the new one to be healthy?

**Comment by liewegas** [comment]
Two things:

- Ideally we should leverage the 'ceph osd ok-to-stop' to know if it's fine to upgrade the next osd.  it's smart enough to say no if stopping the osd would make pgs unable to peer.  There is some risk, though, that a poorly set up cluster will simply be unable to proceed with restarts because the osd can't be stopped, maybe because another osd is down.
- After an osd is started, we probably want a settle() or similar function that will make sure the osd is up and pgs peer and the pg stats that the mgr has are up to date.  Maybe the way to do this is wait_for_up_and_peered(int osd) that waits until the osd comes up, and then does the 'cell tell osd.N flush_pg_stats' + 'ceph osd last-stat-seq osd.NNN' commands (which let you correctly block until the pg stats are reported to the mgr/mon).  See qa/standalone/ceph-helpers.sh flush_pg_stats() for a sample implementation.

**Comment by travisn** [comment]
This was already fixed with #2901
